From 02b703be340cb1ae9775b286011f8745503b975e Mon Sep 17 00:00:00 2001
From: techyminati <sinha.aryan03@gmail.com>
Date: Sun, 21 Jul 2024 22:19:23 +0530
Subject: [PATCH] SystemUI: Re-implement Volla Lockscreen Clock

* Partially based on patch f923bc40e67b70d7cfeffa2e81f73a2902aacacd & forward-ported to AOSP 14 QPR3

Change-Id: Ie0d2a83648d4b338224fc1e3d8f1ee0753528c48
Signed-off-by: Aryan Sinha <techyminati@outlook.com>
---
 .../android/systemui/animation/TextInterpolator.kt   |  9 ++++++++-
 .../customization/res/layout/clock_default_large.xml | 12 +++++++-----
 .../customization/res/layout/clock_default_small.xml |  9 +++++----
 .../res-keyguard/layout/keyguard_clock_switch.xml    |  4 +++-
 4 files changed, 23 insertions(+), 11 deletions(-)

diff --git a/packages/SystemUI/animation/src/com/android/systemui/animation/TextInterpolator.kt b/packages/SystemUI/animation/src/com/android/systemui/animation/TextInterpolator.kt
index 02caeeddd774..a3d10e8ee1c5 100644
--- a/packages/SystemUI/animation/src/com/android/systemui/animation/TextInterpolator.kt
+++ b/packages/SystemUI/animation/src/com/android/systemui/animation/TextInterpolator.kt
@@ -17,6 +17,7 @@ package com.android.systemui.animation
 
 import android.graphics.Canvas
 import android.graphics.Paint
+import android.graphics.Color
 import android.graphics.fonts.Font
 import android.graphics.fonts.FontVariationAxis
 import android.graphics.text.PositionedGlyphs
@@ -245,10 +246,16 @@ class TextInterpolator(
                     // Move to drawing origin.
                     val origin = layout.getDrawOrigin(lineNo)
                     canvas.translate(origin, layout.getLineBaseline(lineNo).toFloat())
-
+                    // techyminati: Implement VollaOS Clock
+                    var oldColor = tmpPaint.color
+                    if (lineNo == 1) {
+                       tmpPaint.setColor(Color.parseColor("#FF3615"))
+                   }
+                
                     run.fontRuns.forEach { fontRun ->
                         drawFontRun(canvas, run, fontRun, lineNo, tmpPaint)
                     }
+                tmpPaint.setColor(oldColor)
                 } finally {
                     canvas.restore()
                 }
diff --git a/packages/SystemUI/customization/res/layout/clock_default_large.xml b/packages/SystemUI/customization/res/layout/clock_default_large.xml
index 0139d50dcfba..86a5f9c83901 100644
--- a/packages/SystemUI/customization/res/layout/clock_default_large.xml
+++ b/packages/SystemUI/customization/res/layout/clock_default_large.xml
@@ -2,6 +2,7 @@
 <!--
 **
 ** Copyright 2022, The Android Open Source Project
+** Copyright 2024, Volla Systeme GmbH
 **
 ** Licensed under the Apache License, Version 2.0 (the "License")
 ** you may not use this file except in compliance with the License.
@@ -20,12 +21,13 @@
     xmlns:android="http://schemas.android.com/apk/res/android"
     android:layout_width="wrap_content"
     android:layout_height="wrap_content"
-    android:layout_gravity="center"
+    android:layout_gravity="center_vertical"
+    android:layout_marginStart="30dp"
     android:gravity="center_horizontal"
     android:textSize="@dimen/large_clock_text_size"
-    android:fontFamily="@*android:string/config_clockFontFamily"
-    android:typeface="monospace"
+    android:fontFamily="ibmplexsans"
+    android:textStyle="bold"
     android:elegantTextHeight="false"
     chargeAnimationDelay="200"
-    dozeWeight="200"
-    lockScreenWeight="400" />
+    dozeWeight="300"
+    lockScreenWeight="900" />
diff --git a/packages/SystemUI/customization/res/layout/clock_default_small.xml b/packages/SystemUI/customization/res/layout/clock_default_small.xml
index ff6d7f9e2240..52dd66b07b80 100644
--- a/packages/SystemUI/customization/res/layout/clock_default_small.xml
+++ b/packages/SystemUI/customization/res/layout/clock_default_small.xml
@@ -23,11 +23,12 @@
     android:layout_gravity="start"
     android:gravity="start"
     android:textSize="@dimen/small_clock_text_size"
-    android:fontFamily="@*android:string/config_clockFontFamily"
+    android:textStyle="bold"
+    android:fontFamily="ibmplexsans"
     android:elegantTextHeight="false"
     android:ellipsize="none"
-    android:singleLine="true"
+    android:singleLine="false"
     android:fontFeatureSettings="pnum"
     chargeAnimationDelay="350"
-    dozeWeight="200"
-    lockScreenWeight="400" />
+    dozeWeight="300"
+    lockScreenWeight="900" />
diff --git a/packages/SystemUI/res-keyguard/layout/keyguard_clock_switch.xml b/packages/SystemUI/res-keyguard/layout/keyguard_clock_switch.xml
index fc9c917c152b..7728792c257d 100644
--- a/packages/SystemUI/res-keyguard/layout/keyguard_clock_switch.xml
+++ b/packages/SystemUI/res-keyguard/layout/keyguard_clock_switch.xml
@@ -28,10 +28,12 @@
     <com.android.keyguard.KeyguardClockFrame
         android:id="@+id/lockscreen_clock_view"
         android:layout_width="wrap_content"
-        android:layout_height="@dimen/small_clock_height"
+        android:layout_height="wrap_content"
         android:layout_alignParentStart="true"
         android:layout_alignParentTop="true"
         android:clipChildren="false"
+        android:fontFamily="ibmplexsans"
+        android:textStyle="bold"
         android:paddingStart="@dimen/clock_padding_start"
         android:visibility="invisible" />
     <com.android.keyguard.KeyguardClockFrame
-- 
2.29.2

